In [ ]:
# Python Version 3.9.13, conda version 22.9.0 (conda -V)

In [ ]:
import pandas as pd # v 1.5.3
import numpy as np # v 1.21.5
import matplotlib.pyplot as plt # v 3.5.2

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
from sklearn.tree import plot_tree
from sklearn import tree
from imblearn.under_sampling import RandomUnderSampler
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
import time

In [21]:
df = pd.read_csv(r"/Users/yelim/Documents/GitHub/DATA-300-Statistical-Machine-Learning-Fall-2023-/Data/company_bankruptcy_data.csv")

### EDA

In [22]:
df.head()

Bankrupt   ROA(C) before interest and depreciation before interest  \
0         1                                           0.370594          
1         1                                           0.464291          
2         1                                           0.426071          
3         1                                           0.399844          
4         1                                           0.465022          

    ROA(A) before interest and % after tax  \
0                                 0.424389   
1                                 0.538214   
2                                 0.499019   
3                                 0.451265   
4                                 0.538432   

    ROA(B) before interest and depreciation after tax  \
0                                           0.405750    
1                                           0.516730    
2                                           0.472295    
3                                           0.457733    
4                                           0.522298    

    Operating Gross Margin   Realized Sales Gross Margin  \
0                 0.601457                      0.601457   
1                 0.610235                      0.610235   
2                 0.601450                      0.601364   
3                 0.583541                      0.583541   
4                 0.598783                      0.598783   

    Operating Profit Rate   Pre-tax net Interest Rate  \
0                0.998969                    0.796887   
1                0.998946                    0.797380   
2                0.998857                    0.796403   
3                0.998700                    0.796967   
4                0.998973                    0.797366   

    After-tax net Interest Rate   Non-industry income and expenditure/revenue  \
0                      0.808809                                      0.302646   
1                      0.809301                                      0.303556   
2                      0.808388                                      0.302035   
3                      0.808966                                      0.303350   
4                      0.809304                                      0.303475   

   ...   Net Income to Total Assets   Total assets to GNP price  \
0  ...                     0.716845                    0.009219   
1  ...                     0.795297                    0.008323   
2  ...                     0.774670                    0.040003   
3  ...                     0.739555                    0.003252   
4  ...                     0.795016                    0.003878   

    No-credit Interval   Gross Profit to Sales  \
0             0.622879                0.601453   
1             0.623652                0.610237   
2             0.623841                0.601449   
3             0.622929                0.583538   
4             0.623521                0.598782   

    Net Income to Stockholder's Equity   Liability to Equity  \
0                             0.827890              0.290202   
1                             0.839969              0.283846   
2                             0.836774              0.290189   
3                             0.834697              0.281721   
4                             0.839973              0.278514   

    Degree of Financial Leverage (DFL)  \
0                             0.026601   
1                             0.264577   
2                             0.026555   
3                             0.026697   
4                             0.024752   

    Interest Coverage Ratio (Interest expense to EBIT)   Net Income Flag  \
0                                           0.564050                   1   
1                                           0.570175                   1   
2                                           0.563706                   1   
3                                           0.564663                   1   
4                              

In [23]:
df.describe()

Bankrupt   ROA(C) before interest and depreciation before interest  \
count  6819.000000                                        6819.000000          
mean      0.032263                                           0.505180          
std       0.176710                                           0.060686          
min       0.000000                                           0.000000          
25%       0.000000                                           0.476527          
50%       0.000000                                           0.502706          
75%       0.000000                                           0.535563          
max       1.000000                                           1.000000          

        ROA(A) before interest and % after tax  \
count                              6819.000000   
mean                                  0.558625   
std                                   0.065620   
min                                   0.000000   
25%                                   0.535543   
50%                                   0.559802   
75%                                   0.589157   
max                                   1.000000   

        ROA(B) before interest and depreciation after tax  \
count                                        6819.000000    
mean                                            0.553589    
std                                             0.061595    
min                                             0.000000    
25%                                             0.527277    
50%                                             0.552278    
75%                                             0.584105    
max                                             1.000000    

        Operating Gross Margin   Realized Sales Gross Margin  \
count              6819.000000                   6819.000000   
mean                  0.607948                      0.607929   
std                   0.016934                      0.016916   
min                   0.000000                      0.000000   
25%                   0.600445                      0.600434   
50%                   0.605997                      0.605976   
75%                   0.613914                      0.613842   
max                   1.000000                      1.000000   

        Operating Profit Rate   Pre-tax net Interest Rate  \
count             6819.000000                 6819.000000   
mean                 0.998755                    0.797190   
std                  0.013010                    0.012869   
min                  0.000000                    0.000000   
25%                  0.998969                    0.797386   
50%                  0.999022                    0.797464   
75%                  0.999095                    0.797579   
max                  1.000000                    1.000000   

        After-tax net Interest Rate  \
count                   6819.000000   
mean                       0.809084   
std                        0.013601   
min                        0.000000   
25%                        0.809312   
50%                        0.809375   
75%                        0.809469   
max                        1.000000   

        Non-industry income and expenditure/revenue  ...  \
count                                   6819.000000  ...   
mean                                       0.303623  ...   
std                                        0.011163  ...   
min                                        0.000000  ...   
25%                                        0.303466  ...   
50%                                        0.303525  ...   
75%                                        0.303585  ...   
max                                        1.000000  ...   

        Net Income to Total Assets   Total assets to GNP price  \
count                  6819.000000                6.819000e+03   
mean                      0.807760                1.862942e+07   
std                       0.040332                3.764501e+08   
min 

In [24]:
df.isna().sum()

Bankrupt                                                    0
 ROA(C) before interest and depreciation before interest    0
 ROA(A) before interest and % after tax                     0
 ROA(B) before interest and depreciation after tax          0
 Operating Gross Margin                                     0
                                                           ..
 Liability to Equity                                        0
 Degree of Financial Leverage (DFL)                         0
 Interest Coverage Ratio (Interest expense to EBIT)         0
 Net Income Flag                                            0
 Equity to Liability                                        0
Length: 96, dtype: int64

In [25]:
df['Bankrupt'].value_counts()

0    6599
1     220
Name: Bankrupt, dtype: int64

In [ ]:
X = df.loc[:, df.columns != 'Bankrupt']
y = df['Bankrupt']

### Undersample the data

In [27]:
undersampler = RandomUnderSampler(sampling_strategy='majority')
X_under, y_under = undersampler.fit_resample(X, y)

### Split the data into a training and testing set.

In [28]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Part 1 - Random Forest

### 1.	Fit the model on the training set using the default parameters and report your findings.(5 points)

In [29]:
rf_default = RandomForestClassifier()
start_time_rf_default = time.time()
rf_default.fit(X_train, y_train)
end_time_rf_default = time.time()

# Evaluate the default model
y_pred_rf_default = rf_default.predict(X_test)
accuracy_rf_default = accuracy_score(y_test, y_pred_rf_default)
report_rf_default = classification_report(y_test, y_pred_rf_default)

# Print findings
print("1. Default Random Forest Model:")
print(f"   Accuracy: {accuracy_rf_default}")
print("   Classification Report:")
print(report_rf_default)
print(f"   Model fitting time: {end_time_rf_default - start_time_rf_default} seconds")


1. Default Random Forest Model:
   Accuracy: 0.967008797653959
   Classification Report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.98      1313
           1       0.75      0.18      0.29        51

    accuracy                           0.97      1364
   macro avg       0.86      0.59      0.63      1364
weighted avg       0.96      0.97      0.96      1364

   Model fitting time: 2.013514995574951 seconds


### 2.	Repeat 1 but change the parameters to different non-default parameters. Evaluate this model on your choice of metrics. Which model do you prefer? (5 points)

In [37]:
rf_nondefault = RandomForestClassifier(criterion='entropy',n_estimators=50, max_depth=10)
start_time_rf_nondefault = time.time()
rf_nondefault.fit(X_train, y_train)
end_time_rf_nondefault = time.time()

# Evaluate the custom model
y_pred_rf_nondefault = rf_nondefault.predict(X_test)
accuracy_rf_nondefault = accuracy_score(y_test, y_pred_rf_nondefault)
report_rf_nondefault = classification_report(y_test, y_pred_rf_nondefault)

# Print findings
print("\n2. Random Forest Mode with Non-default Parameters:")
print(f"   Accuracy: {accuracy_rf_nondefault}")
print("   Classification Report:")
print(report_rf_nondefault)
print(f"   Model fitting time: {end_time_rf_nondefault - start_time_rf_nondefault} seconds")


2. Random Forest Mode with Non-default Parameters:
   Accuracy: 0.969208211143695
   Classification Report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.98      1313
           1       0.91      0.20      0.32        51

    accuracy                           0.97      1364
   macro avg       0.94      0.60      0.65      1364
weighted avg       0.97      0.97      0.96      1364

   Model fitting time: 0.768761157989502 seconds


#### The two models have similar accuracy of 97%, but the model with non-default parameters has a shorter model fitting time, so I prefer the model with non-default parameters.

### 3.	Which model takes longer to fit? 1. Or 2.? Report the CPU time. (2 points)

#### The model with default parameters (1) takes longer to fit. Model 1 takes 2.01 seconds, and model 2 takes 0.77 seconds.

### 4.	Based on your findings above which parameter combinations give you the best results for classification? Would you prefer “Gini” or “entropy” as your splitting metric? (5 points)

#### Based on my findings above, non-default parameters - criterion='entropy',n_estimators=50, max_depth=10 - has the best results for classification. I prefer "entropy" as my splitting metric, as it yields the shortest model fitting time.

### 5.	Choose one of the models from number 1. Or number 2, display, and discuss the decision rules. Do the rules make sense for classification? (5 points)

In [42]:
text_representation = tree.export_text(rf_nondefault.estimators_[0], feature_names=list(df.loc[:, df.columns != 'Bankrupt']))
print(text_representation)

|---  Retained Earnings to Total Assets <= 0.93
|   |---  Current Liabilities/Equity <= 0.34
|   |   |---  Net Income to Stockholder's Equity <= 0.84
|   |   |   |---  Average Collection Days <= 0.00
|   |   |   |   |--- class: 1.0
|   |   |   |---  Average Collection Days >  0.00
|   |   |   |   |---  Persistent EPS in the Last Four Seasons <= 0.18
|   |   |   |   |   |---  Operating Profit Growth Rate <= 0.85
|   |   |   |   |   |   |---  Pre-tax net Interest Rate <= 0.80
|   |   |   |   |   |   |   |---  Allocation rate per person <= 0.04
|   |   |   |   |   |   |   |   |--- class: 0.0
|   |   |   |   |   |   |   |---  Allocation rate per person >  0.04
|   |   |   |   |   |   |   |   |---  Cash/Total Assets <= 0.20
|   |   |   |   |   |   |   |   |   |--- class: 0.0
|   |   |   |   |   |   |   |   |---  Cash/Total Assets >  0.20
|   |   |   |   |   |   |   |   |   |--- class: 1.0
|   |   |   |   |   |   |---  Pre-tax net Interest Rate >  0.80
|   |   |   |   |   |   |   |--- class:

#### Looking at the results, if retained earnings to total assets is greater than 0.93, interest coverage ratio is less than or equal to 0.57, debt ratio % is less than or equal to 0.18. net value growth rate is less than 0.00, debt ratio % is less than or equal to 0.09, the model classifies the observation as 0 indicating no bankruptcy which logically makes sense for classification.

### 6.	Which features are the most important for classification? (2 points)

In [59]:
feature_names=list(df.loc[:, df.columns != 'Bankrupt'])
feature_importance_rf = rf_default.feature_importances_
sorted_features = sorted(zip(feature_names, feature_importance_rf), key=lambda x: x[1], reverse=True)

print("\nTop 5 Feature Importance (Random Forest):")
for feature, importance in sorted_features[:5]:
    print(f"{feature}: {importance}")


Top 5 Feature Importance (Random Forest):
 Net Income to Stockholder's Equity: 0.0310408012332774
 Net Value Growth Rate: 0.0306550616057181
 Net profit before tax/Paid-in capital: 0.0232150578843515
 Borrowing dependency: 0.020303876283580054
 Net Income to Total Assets: 0.01980961157841311


#### The most important features are 

### 7.	On average, what do you think is the tradeoff between model fitting time and model performance?  (3 points)

#### The more complex the model is, it potentially leads to higher predictive performance. However, with increased model complexity often comes with a higher computational cost during training. This is important especially when dealing with large datasets. Therefore, if we want to have a model with better performance, the longer the model fitting time tends to be.

## Part 1 - SVM

### 1.	Fit the model on the training set using the default parameters and report your findings.(5 points)

In [ ]:
svm_default = SVC()
start_time_svm_default = time.time()
svm_default.fit(X_train, y_train)
end_time_svm_default = time.time()

# Evaluate the default SVM model
y_pred_default = svm_default.predict(X_test)
accuracy_default = accuracy_score(y_test, y_pred_default)
report_default = classification_report(y_test, y_pred_default)

# Print findings
print("1. Default SVM Model:")
print(f"   Accuracy: {accuracy_default}")
print("   Classification Report:")
print(report_default)
print(f"   Model fitting time: {end_time_svm_default - start_time_svm_default} seconds")

### 2.	Change the kernel from “rbf” to “linear” and repeat the model fitting procedure in 1 above (5 point)

In [ ]:
svm_linear = SVC(kernel='linear')
start_time_svm_linear = time.time()
svm_linear.fit(X_train, y_train)
end_time_svm_linear = time.time()

# Evaluate the linear SVM model
y_pred_linear = svm_linear.predict(X_test)
accuracy_linear = accuracy_score(y_test, y_pred_linear)
report_linear = classification_report(y_test, y_pred_linear)

# Print findings
print("\n2. Linear SVM Model:")
print(f"   Accuracy: {accuracy_linear}")
print("   Classification Report:")
print(report_linear)
print(f" Model fitting time: {end_time_svm_linear - start_time_svm_linear} seconds")

### 3.	Evaluate both models using any metrics of choice. (4 points)

#### Based on accuracy of both models, since has an accuracy of and has an accuracy of, has better performance.

### 4.	How long does it take to fit the SVM model in 1. above? How about 2. above? Report only the CPU times (1 point)

#### It takes to fit the SVM model in 1, and it takes to fit the SVM model in 2.